### PR TITLE
feat: add lexcial and syntax support for rlike

### PR DESCRIPTION
**Pull request by jiang1997**

### CODE REVIEW

**Comment by jiang1997** [comment]
I tried to add support for rlike, but can't make it pass the parser_is_distinct_test.
The details are shown as follows.
```
//zetasql/parser:parser_whitespace_test                                  PASSED in 0.3s
//zetasql/parser:parser_with_offset_test                                 PASSED in 0.1s
//zetasql/parser:parser_with_test                                        PASSED in 0.1s
//zetasql/parser:unparser_test                                           PASSED in 0.0s
//zetasql/parser:parser_is_distinct_test                                 FAILED in 0.1s
  /home/jieke/.cache/bazel/_bazel_jieke/b3a5130ebede314b1a43d74a26418495/execroot/com_google_zetasql/bazel-out/k8-fastbuild/testlogs/zetasql/parser/parser_is_distinct_test/test.log

Executed 125 out of 125 tests: 124 tests pass and 1 fails locally.
INFO: Build completed, 1 test FAILED, 126 total actions
```
test.log
```
=================== DIFF ===============================
--- expected/zetasql/parser/testdata/is_distinct.test
+++ actual/zetasql/parser/testdata/is_distinct.test
@@ -1,7 +1,7 @@
 
 # Syntax error
 SELECT 1 NOT DISTINCT FROM 2;
 --
-ERROR: Syntax error: Expected keyword BETWEEN or keyword ILIKE or keyword IN or keyword LIKE but got keyword DISTINCT [at 1:14]
+ERROR: Syntax error: Unexpected keyword DISTINCT [at 1:14]
 SELECT 1 NOT DISTINCT FROM 2;
              ^

external/com_google_file_based_test_driver/file_based_test_driver/file_based_test_driver.cc:304: Failure
Failed
```
# What I know
https://github.com/4paradigm/zetasql/blob/3bd67edcfcdf5d59fb945ba9f7d4d5de4d29e78c/zetasql/parser/bison_parser.cc#L156
This function deals with the final error, and if there is no expectation-like in the message from Bison the final error will be like ``Unexpected keyword DISTINCT``. So it must be some modifications affecting the way Bison output message.

Do you have any advice on this issue, please?

**Comment by jiang1997** [comment]
@aceforeverd 
I thought I found the reason, and it's normal for Bison to behave like this.
As the number of tokens could be placed directly behind "NOT" increases, Bison's error message, for some reason, will change from ``unexpected XX, expecting XX or XX or XX or XX`` to ``unexpected XX``.
Specifically, 4 is borderline.

Here is the output from Bison when I increase the token number from 4 to 5.
```
Parse error: syntax error, unexpected T_DISTINCT, expecting T_LIKE or T_l1 or T_l2 or T_l3
Parse error: syntax error, unexpected T_DISTINCT
```

**Comment by aceforeverd** [comment]
> @aceforeverd 
> 
> I thought I found the reason, and it's normal for Bison to behave like this.
> 
> As the number of tokens could be placed directly behind "NOT" increases, Bison's error message, for some reason, will change from ``unexpected XX, expecting XX or XX or XX or XX`` to ``unexpected XX``.
> 
> Specifically, 4 is borderline.
> 
> 
> 
> Here is the output from Bison when I increase the token number from 4 to 5.
> 
> ```
> 
> Parse error: syntax error, unexpected T_DISTINCT, expecting T_LIKE or T_l1 or T_l2 or T_l3
> 
> Parse error: syntax error, unexpected T_DISTINCT
> 
> ```

Bingo, refer the ilike pr 😀

**Comment by aceforeverd** [comment]
thanks, this pr need merged before your PR in OpenMLDB compile

File: zetasql/parser/bison_parser.y [hunk 1]
```diff
@@ -5630,6 +5641,27 @@ expression:
...
+          if (IsUnparenthesizedNotExpression($3)) {
+            YYERROR_UNEXPECTED_AND_ABORT_AT(@3);
+          }
+          // Bison allows some cases like IN on the left hand side because it's
+          // not ambiguous. The language doesn't allow this.
+          if (!$1->IsAllowedInComparison()) {
+            YYERROR_AND_ABORT_AT(
+                @2,
+                "Syntax error: "
+                "Expression to the left of LIKE must be parenthesized");
```
Comment (aceforeverd) [inline zetasql/parser/bison_parser.y]: RLIKE

**Comment by aceforeverd** [comment]
> 

kind annoying, tested and original bison message just go empty

File: zetasql/parser/bison_parser.y (hunk 1 above)
Comment (jiang1997) [reply zetasql/parser/bison_parser.y]: Got it.

**Comment by jiang1997** [comment]
> > 
> 
> kind annoying, tested and original bison message just go empty

It's two pieces of code that control how Bison reports syntax error.
In my test case, I found these from ``calc.tab.c``. But I have not found the counterpart for this project.
```cpp
# define YYCASE_(N, S)                      \
      case N:                               \
        yyformat = S;                       \
      break
    default: /* Avoid compiler warnings. */
      YYCASE_(0, YY_("syntax error"));
      YYCASE_(1, YY_("syntax error, unexpected %s"));
      YYCASE_(2, YY_("syntax error, unexpected %s, expecting %s"));
      YYCASE_(3, YY_("syntax error, unexpected %s, expecting %s or %s"));
      YYCASE_(4, YY_("syntax error, unexpected %s, expecting %s or %s or %s"));
      YYCASE_(5, YY_("syntax error, unexpected %s, expecting %s or %s or %s or %s"));
# undef YYCASE_
    }
```

```cpp
yysyntax_error (YYPTRDIFF_T *yymsg_alloc, char **yymsg,
                yy_state_t *yyssp, int yytoken)
{
  enum { YYERROR_VERBOSE_ARGS_MAXIMUM = 5 };
  /* Internationalized format string. */
  const char *yyformat = YY_NULLPTR;
  /* Arguments of yyformat: reported tokens (one for the "unexpected",
     one per "expected"). */
  char const *yyarg[YYERROR_VERBOSE_ARGS_MAXIMUM];
  /* Actual size of YYARG. */
  int yycount = 0;
  /* Cumulated lengths of YYARG.  */
  YYPTRDIFF_T yysize = 0;

```
It' seems not tricky to make Bison behave consistently when the number of token increase to 5.
Just add another line ``YYCASE_(6, YY_("syntax error, unexpected %s, expecting %s or %s or %s or %s or %s"));`` and change ``YYERROR_VERBOSE_ARGS_MAXIMUM`` from 5 to 6.
It works for my test case.

**Comment by aceforeverd** [comment]
> > > 
> > 
> > 
> > kind annoying, tested and original bison message just go empty
> 
> It's two pieces of code that control how Bison reports syntax error. In my test case, I found these from `calc.tab.c`. But I have not found the counterpart for this project.
> 
> ```c++
> # define YYCASE_(N, S)                      \
>       case N:                               \
>         yyformat = S;                       \
>       break
>     default: /* Avoid compiler warnings. */
>       YYCASE_(0, YY_("syntax error"));
>       YYCASE_(1, YY_("syntax error, unexpected %s"));
>       YYCASE_(2, YY_("syntax error, unexpected %s, expecting %s"));
>       YYCASE_(3, YY_("syntax error, unexpected %s, expecting %s or %s"));
>       YYCASE_(4, YY_("syntax error, unexpected %s, expecting %s or %s or %s"));
>       YYCASE_(5, YY_("syntax error, unexpected %s, expecting %s or %s or %s or %s"));
> # undef YYCASE_
>     }
> ```
> 
> ```c++
> yysyntax_error (YYPTRDIFF_T *yymsg_alloc, char **yymsg,
>                 yy_state_t *yyssp, int yytoken)
> {
>   enum { YYERROR_VERBOSE_ARGS_MAXIMUM = 5 };
>   /* Internationalized format string. */
>   const char *yyformat = YY_NULLPTR;
>   /* Arguments of yyformat: reported tokens (one for the "unexpected",
>      one per "expected"). */
>   char const *yyarg[YYERROR_VERBOSE_ARGS_MAXIMUM];
>   /* Actual size of YYARG. */
>   int yycount = 0;
>   /* Cumulated lengths of YYARG.  */
>   YYPTRDIFF_T yysize = 0;
> ```
> 
> It' seems not tricky to make Bison behave consistently when the number of token increase to 5. Just add another line `YYCASE_(6, YY_("syntax error, unexpected %s, expecting %s or %s or %s or %s or %s"));` and change `YYERROR_VERBOSE_ARGS_MAXIMUM` from 5 to 6. It works for my test case.

Nice. So it's limit from bison source and no option to tweak it, we're not going to fix it anyway.

**Comment by aceforeverd** [comment]
@jiang1997  I'm ok for this PR . Do u have any more changes, if not maybe u can change state to `ready for review`, so I can merge it.  Have to update thirdparty version in OpenMLDB PR in order to pass CI tests

**Comment by jiang1997** [comment]
> @jiang1997 I'm ok for this PR . Do u have any more changes, if not maybe u can change state to `ready for review`, so I can merge it. Have to update thirdparty version in OpenMLDB PR in order to pass CI tests

can't wait!

**Review by aceforeverd — APPROVED** [review]

**Comment by aceforeverd** [comment]
@jiang1997 sorry last thing, resign commits with gpg sign

**Comment by aceforeverd** [comment]
merged. @jiang1997 you should update `third-party/cmake/FetchZetasql.cmake` to the latest version in openmldb

**Comment by jiang1997** [comment]
> merged. @jiang1997 you should update `third-party/cmake/FetchZetasql.cmake` to the latest version in openmldb

Uh, you mean I am supposed to create another pr to apply this feat to openmldb?

**Comment by aceforeverd** [comment]
> > merged. @jiang1997 you should update `third-party/cmake/FetchZetasql.cmake` to the latest version in openmldb
> 
> Uh, you mean I am supposed to create another pr to apply this feat to openmldb?

Nop, just here https://github.com/4paradigm/OpenMLDB/pull/2187. This is required for CI to pass

**Comment by jiang1997** [comment]
> > > merged. @jiang1997 you should update `third-party/cmake/FetchZetasql.cmake` to the latest version in openmldb
> > 
> > 
> > Uh, you mean I am supposed to create another pr to apply this feat to openmldb?
> 
> Nop, just here [4paradigm/OpenMLDB#2187](https://github.com/4paradigm/OpenMLDB/pull/2187). This is required for CI to pass

Got it!

**Pull request merged**